File: lib/prefetch.c
//
// Prefetch Implementation - CUDA Only
// Manages concurrent tensor loading with async GPU operations.
//
// Note: CPU prefetch is handled directly in Rust (mmap is lazy, no async needed).
//

#include "hmll/prefetch.h"
#include "hmll/hmll.h"

#include <stdlib.h>
#include <string.h>

#ifdef __HMLL_CUDA_ENABLED__

#include "hmll/cuda_pool.h"
#include <cuda_runtime_api.h>

struct hmll_error hmll_prefetch_init(
    struct hmll_prefetch_ctx* ctx,
    size_t num_slots,
    enum hmll_device device,
    int device_id
) {
    if (!ctx) {
        return HMLL_ERR(HMLL_ERR_ALLOCATION_FAILED);
    }

    // Only CUDA is supported
    if (device != HMLL_DEVICE_CUDA) {
        return HMLL_ERR(HMLL_ERR_INVALID_DEVICE);
    }

    memset(ctx, 0, sizeof(*ctx));

    // Clamp slot count
    if (num_slots == 0) num_slots = HMLL_PREFETCH_DEFAULT_SLOTS;
    if (num_slots > HMLL_PREFETCH_MAX_SLOTS) num_slots = HMLL_PREFETCH_MAX_SLOTS;

    ctx->device = device;
    ctx->device_id = device_id;
    ctx->num_slots = num_slots;
    ctx->next_slot = 0;

    ctx->slots = calloc(num_slots, sizeof(struct hmll_prefetch_slot));
    if (!ctx->slots) {
        return HMLL_ERR(HMLL_ERR_ALLOCATION_FAILED);
    }

    // Initialize each slot with its own stream and event
    for (size_t i = 0; i < num_slots; i++) {
        struct hmll_prefetch_slot* slot = &ctx->slots[i];
        slot->state = HMLL_PREFETCH_IDLE;
        slot->tensor_index = 0;
        slot->buffer.ptr = NULL;
        slot->buffer.size = 0;
        slot->buffer.device = HMLL_DEVICE_CUDA;

        cudaError_t err = cudaSetDevice(device_id);
        if (err != cudaSuccess) {
            goto cleanup;
        }

        cudaStream_t stream;
        err = cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking);
        if (err != cudaSuccess) {
            goto cleanup;
        }
        slot->stream = stream;

        cudaEvent_t event;
        err = cudaEventCreateWithFlags(&event, cudaEventDisableTiming);
        if (err != cudaSuccess) {
            cudaStreamDestroy(stream);
            slot->stream = NULL;
            goto cleanup;
        }
        slot->done_event = event;
        continue;

    cleanup:
        // Cleanup already-created resources
        for (size_t j = 0; j < i; j++) {
            if (ctx->slots[j].stream) {
                cudaStreamDestroy((cudaStream_t)ctx->slots[j].stream);
            }
            if (ctx->slots[j].done_event) {
                cudaEventDestroy((cudaEvent_t)ctx->slots[j].done_event);
            }
        }
        free(ctx->slots);
        ctx->slots = NULL;
        return HMLL_ERR(HMLL_ERR_CUDA_ERROR);
    }

    // Initialize CUDA memory pool
    struct hmll_error pool_err = hmll_cuda_pool_init(device_id);
    if (hmll_check(pool_err)) {
        hmll_prefetch_destroy(ctx);
        return pool_err;
    }

    return HMLL_OK;
}

struct hmll_error hmll_prefetch_start_load(
    struct hmll_prefetch_ctx* ctx,
    const void* src_ptr,
    size_t size,
    size_t tensor_index,
    size_t* out_slot
) {
    if (!ctx || !ctx->slots) {
        return HMLL_ERR(HMLL_ERR_ALLOCATION_FAILED);
    }

    int slot_idx = hmll_prefetch_find_available_slot(ctx);
    if (slot_idx < 0) {
        return HMLL_ERR(HMLL_ERR_BUFFER_TOO_SMALL);
    }

    struct hmll_prefetch_slot* slot = &ctx->slots[slot_idx];

    // Handle zero-size tensors
    if (size == 0 || src_ptr == NULL) {
        slot->buffer.ptr = NULL;
        slot->buffer.size = 0;
        slot->tensor_index = tensor_index;
        slot->state = HMLL_PREFETCH_READY;
        if (out_slot) *out_slot = (size_t)slot_idx;
        return HMLL_OK;
    }

    slot->tensor_index = tensor_index;
    slot->state = HMLL_PREFETCH_LOADING;

    cudaStream_t stream = (cudaStream_t)slot->stream;
    cudaEvent_t event = (cudaEvent_t)slot->done_event;

    // Allocate from pool
    void* ptr = hmll_cuda_pool_alloc_async(size, stream, ctx->device_id);
    if (!ptr) {
        slot->state = HMLL_PREFETCH_ERROR;
        return HMLL_ERR(HMLL_ERR_ALLOCATION_FAILED);
    }

    slot->buffer.ptr = ptr;
    slot->buffer.size = size;
    slot->buffer.device = HMLL_DEVICE_CUDA;

    // Async copy
    cudaError_t err = cudaMemcpyAsync(ptr, src_ptr, size, cudaMemcpyHostToDevice, stream);
    if (err != cudaSuccess) {
        hmll_cuda_pool_free_async(ptr, stream);
        slot->buffer.ptr = NULL;
        slot->state = HMLL_PREFETCH_ERROR;
        return HMLL_ERR(HMLL_ERR_CUDA_ERROR);
    }

    // Record completion
    err = cudaEventRecord(event, stream);
    if (err != cudaSuccess) {
        hmll_cuda_pool_free_async(ptr, stream);
        slot->buffer.ptr = NULL;
        slot->state = HMLL_PREFETCH_ERROR;
        return HMLL_ERR(HMLL_ERR_CUDA_ERROR);
    }

    if (out_slot) *out_slot = (size_t)slot_idx;
    return HMLL_OK;
}

int hmll_prefetch_find_available_slot(struct hmll_prefetch_ctx* ctx) {
    if (!ctx || !ctx->slots) return -1;

    hmll_prefetch_poll(ctx);

    for (size_t i = 0; i < ctx->num_slots; i++) {
        size_t idx = (ctx->next_slot + i) % ctx->num_slots;
        if (ctx->slots[idx].state == HMLL_PREFETCH_IDLE) {
            ctx->next_slot = (idx + 1) % ctx->num_slots;
            return (int)idx;
        }
    }

    return -1;
}

int hmll_prefetch_slot_ready(struct hmll_prefetch_ctx* ctx, size_t slot_index) {
    if (!ctx || !ctx->slots || slot_index >= ctx->num_slots) {
        return 0;
    }

    struct hmll_prefetch_slot* slot = &ctx->slots[slot_index];

    if (slot->state == HMLL_PREFETCH_READY || slot->state == HMLL_PREFETCH_ERROR) {
        return 1;
    }

    if (slot->state != HMLL_PREFETCH_LOADING) {
        return 0;
    }

    cudaError_t err = cudaEventQuery((cudaEvent_t)slot->done_event);
    if (err == cudaSuccess) {
        slot->state = HMLL_PREFETCH_READY;
        return 1;
    } else if (err == cudaErrorNotReady) {
        return 0;
    } else {
        slot->state = HMLL_PREFETCH_ERROR;
        return 1;
    }
}

struct hmll_error hmll_prefetch_wait_slot(struct hmll_prefetch_ctx* ctx, size_t slot_index) {
    if (!ctx || !ctx->slots || slot_index >= ctx->num_slots) {
        return HMLL_ERR(HMLL_ERR_INVALID_RANGE);
    }

    struct hmll_prefetch_slot* slot = &ctx->slots[slot_index];

    if (slot->state == HMLL_PREFETCH_READY) {
        return HMLL_OK;
    }
    if (slot->state == HMLL_PREFETCH_ERROR) {
        return HMLL_ERR(HMLL_ERR_IO_ERROR);
    }
    if (slot->state != HMLL_PREFETCH_LOADING) {
        return HMLL_ERR(HMLL_ERR_INVALID_RANGE);
    }

    cudaError_t err = cudaEventSynchronize((cudaEvent_t)slot->done_event);
    if (err != cudaSuccess) {
        slot->state = HMLL_PREFETCH_ERROR;
        return HMLL_ERR(HMLL_ERR_CUDA_ERROR);
    }

    slot->state = HMLL_PREFETCH_READY;
    return HMLL_OK;
}

int hmll_prefetch_find_tensor(struct hmll_prefetch_ctx* ctx, size_t tensor_index) {
    if (!ctx || !ctx->slots) return -1;

    for (size_t i = 0; i < ctx->num_slots; i++) {
        struct hmll_prefetch_slot* slot = &ctx->slots[i];
        if ((slot->state == HMLL_PREFETCH_LOADING || slot->state == HMLL_PREFETCH_READY) &&
            slot->tensor_index == tensor_index) {
            return (int)i;
        }
    }

    return -1;
}

struct hmll_error hmll_prefetch_take_buffer(
    struct hmll_prefetch_ctx* ctx,
    size_t slot_index,
    struct hmll_iobuf* out_buffer
) {
    if (!ctx || !ctx->slots || slot_index >= ctx->num_slots || !out_buffer) {
        return HMLL_ERR(HMLL_ERR_INVALID_RANGE);
    }

    struct hmll_prefetch_slot* slot = &ctx->slots[slot_index];

    if (slot->state == HMLL_PREFETCH_LOADING) {
        struct hmll_error err = hmll_prefetch_wait_slot(ctx, slot_index);
        if (hmll_check(err)) return err;
    }

    if (slot->state == HMLL_PREFETCH_ERROR) {
        slot->state = HMLL_PREFETCH_IDLE;
        return HMLL_ERR(HMLL_ERR_IO_ERROR);
    }

    if (slot->state != HMLL_PREFETCH_READY) {
        return HMLL_ERR(HMLL_ERR_INVALID_RANGE);
    }

    *out_buffer = slot->buffer;
    slot->buffer.ptr = NULL;
    slot->buffer.size = 0;
    slot->state = HMLL_PREFETCH_IDLE;

    return HMLL_OK;
}

void hmll_prefetch_poll(struct hmll_prefetch_ctx* ctx) {
    if (!ctx || !ctx->slots) return;

    for (size_t i = 0; i < ctx->num_slots; i++) {
        struct hmll_prefetch_slot* slot = &ctx->slots[i];
        if (slot->state == HMLL_PREFETCH_LOADING) {
            cudaError_t err = cudaEventQuery((cudaEvent_t)slot->done_event);
            if (err == cudaSuccess) {
                slot->state = HMLL_PREFETCH_READY;
            } else if (err != cudaErrorNotReady) {
                slot->state = HMLL_PREFETCH_ERROR;
            }
        }
    }
}

void hmll_prefetch_destroy(struct hmll_prefetch_ctx* ctx) {
    if (!ctx || !ctx->slots) return;

    for (size_t i = 0; i < ctx->num_slots; i++) {
        struct hmll_prefetch_slot* slot = &ctx->slots[i];
        cudaStream_t stream = (cudaStream_t)slot->stream;

        if (slot->buffer.ptr) {
            hmll_cuda_pool_free_async(slot->buffer.ptr, stream);
        }

        if (stream) {
            cudaStreamSynchronize(stream);
            cudaStreamDestroy(stream);
        }

        if (slot->done_event) {
            cudaEventDestroy((cudaEvent_t)slot->done_event);
        }
    }

    free(ctx->slots);
    ctx->slots = NULL;
    ctx->num_slots = 0;
}

#else // !__HMLL_CUDA_ENABLED__

// Stub implementations for non-CUDA builds
// These return errors since prefetch is CUDA-only

struct hmll_error hmll_prefetch_init(
    struct hmll_prefetch_ctx* ctx,
    size_t num_slots,
    enum hmll_device device,
    int device_id
) {
    HMLL_UNUSED(ctx);
    HMLL_UNUSED(num_slots);
    HMLL_UNUSED(device);
    HMLL_UNUSED(device_id);
    return HMLL_ERR(HMLL_ERR_CUDA_NOT_ENABLED);
}

struct hmll_error hmll_prefetch_start_load(
    struct hmll_prefetch_ctx* ctx,
    const void* src_ptr,
    size_t size,
    size_t tensor_index,
    size_t* out_slot
) {
    HMLL_UNUSED(ctx);
    HMLL_UNUSED(src_ptr);
    HMLL_UNUSED(size);
    HMLL_UNUSED(tensor_index);
    HMLL_UNUSED(out_slot);
    return HMLL_ERR(HMLL_ERR_CUDA_NOT_ENABLED);
}

int hmll_prefetch_find_available_slot(struct hmll_prefetch_ctx* ctx) {
    HMLL_UNUSED(ctx);
    return -1;
}

int hmll_prefetch_slot_ready(struct hmll_prefetch_ctx* ctx, size_t slot_index) {
    HMLL_UNUSED(ctx);
    HMLL_UNUSED(slot_index);
    return 0;
}

struct hmll_error hmll_prefetch_wait_slot(struct hmll_prefetch_ctx* ctx, size_t slot_index) {
    HMLL_UNUSED(ctx);
    HMLL_UNUSED(slot_index);
    return HMLL_ERR(HMLL_ERR_CUDA_NOT_ENABLED);
}

int hmll_prefetch_find_tensor(struct hmll_prefetch_ctx* ctx, size_t tensor_index) {
    HMLL_UNUSED(ctx);
    HMLL_UNUSED(tensor_index);
    return -1;
}

struct hmll_error hmll_prefetch_take_buffer(
    struct hmll_prefetch_ctx* ctx,
    size_t slot_index,
    struct hmll_iobuf* out_buffer
) {
    HMLL_UNUSED(ctx);
    HMLL_UNUSED(slot_index);
    HMLL_UNUSED(out_buffer);
    return HMLL_ERR(HMLL_ERR_CUDA_NOT_ENABLED);
}

void hmll_prefetch_poll(struct hmll_prefetch_ctx* ctx) {
    HMLL_UNUSED(ctx);
}

void hmll_prefetch_destroy(struct hmll_prefetch_ctx* ctx) {
    HMLL_UNUSED(ctx);
}

#endif // __HMLL_CUDA_ENABLED__

File: lib/rust/hmll/src/prefetch.rs
//! Prefetch context for pipelined tensor loading.
//!
//! This module provides async prefetching of tensors, allowing I/O to overlap
//! with user processing (e.g., quantization).

use crate::{Buffer, Device, Error, Result};
use std::ptr;

/// State of a prefetch slot.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SlotState {
    /// Slot is idle and available.
    Idle,
    /// Copying to pinned staging buffer (sync phase).
    Staging,
    /// Async H2D load in progress.
    Loading,
    /// Load complete, buffer ready.
    Ready,
    /// Load failed.
    Error,
}

impl From<hmll_sys::hmll_prefetch_state> for SlotState {
    fn from(state: hmll_sys::hmll_prefetch_state) -> Self {
        match state {
            hmll_sys::HMLL_PREFETCH_IDLE => SlotState::Idle,
            hmll_sys::HMLL_PREFETCH_STAGING => SlotState::Staging,
            hmll_sys::HMLL_PREFETCH_LOADING => SlotState::Loading,
            hmll_sys::HMLL_PREFETCH_READY => SlotState::Ready,
            hmll_sys::HMLL_PREFETCH_ERROR => SlotState::Error,
            _ => SlotState::Error,
        }
    }
}

/// Prefetch context for managing concurrent tensor loading.
///
/// This context manages multiple "slots" that can load tensors concurrently.
/// For CUDA devices, each slot has its own stream and event for async operations.
///
/// # Example
///
/// ```no_run
/// use hmll::{Device, PrefetchContext};
///
/// # fn main() -> Result<(), hmll::Error> {
/// // Create prefetch context with 4 slots for CUDA device 0
/// let mut ctx = PrefetchContext::new(4, Device::Cuda, 0)?;
///
/// // Start loading tensor data asynchronously
/// let src_data: &[u8] = &[1, 2, 3, 4];
/// let slot = ctx.start_load(src_data.as_ptr() as *const _, src_data.len(), 0)?;
///
/// // ... do other work while loading ...
///
/// // Wait and get the buffer
/// let buffer = ctx.take_buffer(slot)?;
/// # Ok(())
/// # }
/// ```
pub struct PrefetchContext {
    inner: Box<hmll_sys::hmll_prefetch_ctx>,
    device: Device,
}

impl PrefetchContext {
    /// Create a new prefetch context.
    ///
    /// # Arguments
    ///
    /// * `num_slots` - Number of concurrent load slots (1-16)
    /// * `device` - Target device (CPU or CUDA)
    /// * `device_id` - GPU device index (ignored for CPU)
    pub fn new(num_slots: usize, device: Device, device_id: i32) -> Result<Self> {
        let mut inner = Box::new(hmll_sys::hmll_prefetch_ctx {
            slots: ptr::null_mut(),
            num_slots: 0,
            next_slot: 0,
            device_id: 0,
            device: device.to_raw(),
            use_pinned: 1, // Enable pinned memory by default
        });

        unsafe {
            let err = hmll_sys::hmll_prefetch_init(
                inner.as_mut(),
                num_slots,
                device.to_raw(),
                device_id,
            );
            Error::check_hmll_error(err)?;
        }

        Ok(Self { inner, device })
    }

    /// Start async load of tensor data into a slot.
    ///
    /// For CUDA, this allocates GPU memory and starts an async memcpy.
    /// For CPU, this just stores the pointer (zero-copy mmap view).
    ///
    /// # Arguments
    ///
    /// * `src_ptr` - Source data pointer (typically from mmap)
    /// * `size` - Size of tensor data in bytes
    /// * `tensor_index` - Index to identify this tensor
    ///
    /// # Returns
    ///
    /// The slot index that was used.
    pub fn start_load(
        &mut self,
        src_ptr: *const std::ffi::c_void,
        size: usize,
        tensor_index: usize,
    ) -> Result<usize> {
        let mut out_slot: usize = 0;

        unsafe {
            let err = hmll_sys::hmll_prefetch_start_load(
                self.inner.as_mut(),
                src_ptr,
                size,
                tensor_index,
                &mut out_slot,
            );
            Error::check_hmll_error(err)?;
        }

        Ok(out_slot)
    }

    /// Find a slot that is idle or has completed loading.
    ///
    /// Returns `None` if no slot is available.
    pub fn find_available_slot(&mut self) -> Option<usize> {
        let slot = unsafe { hmll_sys::hmll_prefetch_find_available_slot(self.inner.as_mut()) };
        if slot < 0 {
            None
        } else {
            Some(slot as usize)
        }
    }

    /// Check if a specific slot has completed loading (non-blocking).
    pub fn slot_ready(&mut self, slot_index: usize) -> bool {
        unsafe { hmll_sys::hmll_prefetch_slot_ready(self.inner.as_mut(), slot_index) != 0 }
    }

    /// Wait for a specific slot to complete loading (blocking).
    pub fn wait_slot(&mut self, slot_index: usize) -> Result<()> {
        unsafe {
            let err = hmll_sys::hmll_prefetch_wait_slot(self.inner.as_mut(), slot_index);
            Error::check_hmll_error(err)?;
        }
        Ok(())
    }

    /// Find slot containing a specific tensor (by index).
    ///
    /// Returns `None` if the tensor is not in any slot.
    pub fn find_tensor(&mut self, tensor_index: usize) -> Option<usize> {
        let slot = unsafe { hmll_sys::hmll_prefetch_find_tensor(self.inner.as_mut(), tensor_index) };
        if slot < 0 {
            None
        } else {
            Some(slot as usize)
        }
    }

    /// Get buffer from slot, transferring ownership to caller.
    ///
    /// The slot becomes idle after this call.
    /// For CUDA, this waits for the async copy to complete if needed.
    pub fn take_buffer(&mut self, slot_index: usize) -> Result<Buffer> {
        let mut iobuf = hmll_sys::hmll_iobuf {
            ptr: ptr::null_mut(),
            size: 0,
            device: self.device.to_raw(),
        };

        unsafe {
            let err =
                hmll_sys::hmll_prefetch_take_buffer(self.inner.as_mut(), slot_index, &mut iobuf);
            Error::check_hmll_error(err)?;
        }

        // PrefetchContext is CUDA-only; CPU prefetch is handled in Rust (safetensors)
        debug_assert_eq!(self.device, Device::Cuda, "PrefetchContext only supports CUDA");
        Ok(unsafe { Buffer::from_raw_owned(iobuf) })
    }

    /// Poll all slots for completion (updates slot states).
    ///
    /// Call periodically to detect completed loads without blocking.
    pub fn poll(&mut self) {
        unsafe {
            hmll_sys::hmll_prefetch_poll(self.inner.as_mut());
        }
    }

    /// Get the number of slots.
    #[inline]
    pub fn num_slots(&self) -> usize {
        self.inner.num_slots
    }

    /// Get the device type.
    #[inline]
    pub fn device(&self) -> Device {
        self.device
    }

    /// Get state of a specific slot.
    pub fn slot_state(&self, slot_index: usize) -> Option<SlotState> {
        if slot_index >= self.inner.num_slots {
            return None;
        }

        unsafe {
            let slot = &*self.inner.slots.add(slot_index);
            Some(SlotState::from(slot.state))
        }
    }
}

impl Drop for PrefetchContext {
    fn drop(&mut self) {
        unsafe {
            hmll_sys::hmll_prefetch_destroy(self.inner.as_mut());
        }
    }
}

// SAFETY: PrefetchContext can be moved between threads.
// - CUDA streams/events can be used from any thread (not concurrently)
// - All methods require `&mut self`, ensuring exclusive access
// - No `Sync` impl means &PrefetchContext cannot be shared across threads
// These constraints guarantee only one thread accesses the CUDA resources at a time.
unsafe impl Send for PrefetchContext {}

// Tests require CUDA since PrefetchContext is CUDA-only.
// CPU prefetch is handled directly in Rust (safetensors) without going through C.
#[cfg(all(test, feature = "cuda"))]
mod tests {
    use super::*;

    #[test]
    fn test_prefetch_context_cuda() {
        // This test requires an actual CUDA device
        let ctx = PrefetchContext::new(4, Device::Cuda, 0);
        if ctx.is_err() {
            // No CUDA device available, skip test
            return;
        }

        let ctx = ctx.unwrap();
        assert_eq!(ctx.num_slots(), 4);
        assert_eq!(ctx.device(), Device::Cuda);
    }

    #[test]
    fn test_prefetch_cpu_returns_error() {
        // CPU is not supported, should return error
        let ctx = PrefetchContext::new(4, Device::Cpu, 0);
        assert!(ctx.is_err());
    }
}
